src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -227,12 +289,56 @@ public virtual string CreateToken(string payload, SigningCredentials signingCred
+        /// <exception cref="ArgumentNullException">if <paramref name="compressionAlgorithm"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="additionalHeaderClaims"/> is null.</exception>
+        /// <returns>A JWE in compact serialization format.</returns>
+        public virtual string CreateToken(string payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string compressionAlgorithm, IDictionary<string, object> additionalHeaderClaims)
+        {
+            if (string.IsNullOrEmpty(payload))
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            if (signingCredentials == null)
+                throw LogHelper.LogArgumentNullException(nameof(signingCredentials));
+
+            if (encryptingCredentials == null)
+                throw LogHelper.LogArgumentNullException(nameof(encryptingCredentials));
+
+            if (string.IsNullOrEmpty(compressionAlgorithm))
+                throw LogHelper.LogArgumentNullException(nameof(compressionAlgorithm));
+
+            if (additionalHeaderClaims == null)
+                throw LogHelper.LogArgumentNullException(nameof(additionalHeaderClaims));
+
+            return CreateTokenPrivate(JObject.Parse(payload), signingCredentials, encryptingCredentials, compressionAlgorithm, additionalHeaderClaims);
         }
 
-        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string algorithm)
+        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string algorithm, IDictionary<string, object> additionalHeaderClaims)
         {
-            if (!JsonWebTokenManager.KeyToHeaderCache.TryGetValue(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), out string rawHeader))
+
+            if (additionalHeaderClaims != null && (additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Alg)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Kid)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Typ)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.X5t)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Enc)
+                || additionalHeaderClaims.ContainsKey(JwtHeaderParameterNames.Zip)))
+                throw LogHelper.LogExceptionMessage(new SecurityTokenException(LogHelper.FormatInvariant(LogMessages.IDX14116, nameof(additionalHeaderClaims))));
+
+            // if 'encryptingCredentials' != null it means additionalClaims should be added to the outer JWT header.
+            if (encryptingCredentials == null && additionalHeaderClaims != null || !JsonWebTokenManager.KeyToHeaderCache.TryGetValue(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), out string rawHeader))
```

Comment:
This condition looks too complicated. Perhaps we should refactor it.

There might be one more issue here. Let's say that we are creating a JWE token with some additional header claims.
The expected behavior is to add these additional claims only to the outer token header.
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/0f506ac2ef4e7a2a26ddec4369ae4cbe9dfeebcd/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L243

`encryptingCredentials` wouldn't be null, `additionalHeaderClaims` wouldn't be null and let's suppose that the header cache is empty => (false && true) || true => true, so this condition will execute.
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/0f506ac2ef4e7a2a26ddec4369ae4cbe9dfeebcd/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L341

Additional header claims will be added to the inner token header too, which is not in line with the expected behavior. 
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/0f506ac2ef4e7a2a26ddec4369ae4cbe9dfeebcd/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L353-L354

Reply:
Thanks for noticing this! I fixed this bug and added in tests to make sure the additional header claims aren't added to both the inner and the outer JWT headers.

I tried to refactor the logic to make it clearer (and added in some comments), but if you have any other suggestions for improvement let me know! I rebased against the dev branch and the combination of allowing unsigned JWTs and additional header claims has made this method somewhat confusing in my opinion.